mapillm/exact_matching.py

```diff
@@ -0,0 +1,131 @@
+import os
+from mp_api.client import MPRester
+import sys
+
+
+class SuppressOutput:
+    #this is just because I'm annoyed with the print statements and progress bars from mp-api
+    def __enter__(self):
+        self._original_stdout = sys.stdout
+        sys.stdout = open(os.devnull, 'w')
+        self._original_stderr = sys.stderr
+        sys.stderr = open(os.devnull, 'w')
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        sys.stdout.close()
+        sys.stdout = self._original_stdout
+        sys.stderr.close()
+        sys.stderr = self._original_stderr
+
+class ExactMatchingTool():
+  #move to other class when ready
+    def __init__(self):
+        self.mpr = MPRester(os.getenv("MAPI_API_KEY"))
+
+    def _get_specific_query(self, docs, space_group_num=None, density=None, lattice=None):
+        if space_group_num:
+            docs = [doc for doc in docs if doc.symmetry.number == space_group_num]
+        if density:
+            #instead of exact matching density, get density match to the first decimal
+            docs = [doc for doc in docs if round(doc.density, 1) == round(density, 1)]
+        if lattice:
+            docs = [doc for doc in docs if doc.symmetry.crystal_system.value == lattice]
+        return docs
+  
+    def query_material_property(self, formula, desired_prop, space_group_num=None, density=None, lattice=None, one_only=True):
+        with SuppressOutput(): 
+            docs = self.mpr.materials.summary.search(formula=formula, fields=["formula_pretty", desired_prop, "symmetry", "density", "lattice"])
+        docs = self._get_specific_query(docs, space_group_num, density, lattice)
+        if not docs:
+            return []
+        if one_only:
+            docs = [docs[0]]
+        return docs
+  
+    def _get_prop(self, docs, prop):
+        props = []
+        for doc in docs:
+            props.append(getattr(doc, prop))
+        return props
+  
+    def average_neighbors(self, neighbors:dict):
+        total_weight = 0
+        weighted_sum = 0
+        
+        for score, values in neighbors.items():
+            weighted_sum += score * sum(values)
+            total_weight += score * len(values)
+        
+        if total_weight == 0:
+            return 0  
+        else:
+            return weighted_sum / total_weight
+      
+    def average_vanilla(self, all_values:list, property:str):
+        all_values = self._get_prop(all_values, property)
+        return sum(all_values) / len(all_values)
+   
+    def k_vals(self, input_param:dict, k:int):
```

Comment:
I'd change this fxn name for something like `has_k_vals`. It took me a while to understand it was checking if there's enough examples to fill the context already